spec/index.md

```diff
@@ -4474,9 +4510,94 @@ S with
+      let Mod = S.canisters[Canister_id].module
+      let Cert <- { Cert | verify_cert(Cert) and
+                           lookup(["canister", Canister_id, "certified_data"], Cert) = Found S.certified_data[Canister_id] and
+                           lookup(["time"], Cert) = Found S.system_time // or "recent enough"
+                  }
+      if Canister_id ≠ Root_canister_id
+      then
+        Cert := NoCertificate // no certificate available in query and composite query methods evaluated on canisters other than the target canister of the query call
+      let Env = { time = S.time[Canister_id];
+                  global_timer = S.global_timer[Canister_id];
+                  balance = S.balances[Canister_id];
+                  freezing_limit = freezing_limit(S, Canister_id);
+                  certificate = Cert;
+                  status = simple_status(S.canister_status[Canister_id]);
+                  canister_version = S.canister_version[Canister_id];
+                }
+      if S.canisters[Canister_id] ≠ EmptyCanister and
+         S.canister_status[Canister_id] = Running and
+         S.balances[Canister_id] >= freezing_limit(S, Canister_id) and
+         (Method_name ∈ dom(Mod.query_methods) or Method_name ∈ dom(Mod.composite_query_methods)) and
+         Cycles >= MAX_CYCLES_PER_MESSAGE
+      then
+         let W = S.canisters[Canister_id].wasm_state
+         let F = if Method_name ∈ dom(Mod.query_methods) then Mod.query_methods[Method_name] else Mod.composite_query_methods[Method_name]
+         let R = F(Arg, Caller, Env)(W)
+         if R = Trap trap
+         then Return (Reject (CANISTER_ERROR, <implementation-specific>), Cycles - trap.cycles_used)
+         else if R = Return {new_state = W'; new_calls = Calls; response = Response; cycles_used = Cycles_used}
+         then
+            W := W'
+            if Cycles_used > MAX_CYCLES_PER_MESSAGE
+            then
+               Return (Reject (CANISTER_ERROR, <implementation-specific>), Cycles - MAX_CYCLES_PER_MESSAGE) // single message execution out of cycles
+            Cycles := Cycles - Cycles_used
+            if Response = NoResponse
+            then
+               while Calls ≠ []
+               do
+                  if Depth = MAX_CALL_DEPTH_COMPOSITE_QUERY
+                  then
+                     Return (Reject (CANISTER_ERROR, <implementation-specific>), Cycles) // max call graph depth exceeded
+                  let Calls' · Call · Calls''  = Calls
+                  Calls := Calls' · Calls''
+                  if S.canister_subnet[Canister_id] ≠ S.canister_subnet[Call.callee]
+                  then
+                     Return (Reject (CANISTER_ERROR, <implementation-specific>), Cycles) // calling to another subnet
+                  let (Response', Cycles') = composite_query_helper(S, Cycles, Depth + 1, Root_canister_id, Canister_id, Call.callee, Call.method_name, Call.arg)
```

Comment:
I guess that the use of recursion with original S, not updated canister state, ensures that re-entrant calls see the original canister state, and not the transient changes from the outer call, right?

Reply:
Indeed.